misc-utils/waitpid.c

```diff
@@ -0,0 +1,226 @@
+
+static pid_t *parse_pids(size_t n_strings, char * const *strings)
+{
+	pid_t *pids = calloc(n_strings, sizeof(*pids));
+	if (!pids)
+		return NULL;
+
+	for (size_t i = 0; i < n_strings; i++) {
+		char *endptr;
+		const char *arg = strings[i];
+		long long pid = strtoll(arg, &endptr, 10);
+		if (!endptr || pid <= 0) {
+			warnx("invalid PID: '%s'", arg);
+			goto fail;
+		}
+		pids[i] = pid;
+	}
+
+	return pids;
+
+fail:
+	free(pids);
+	return NULL;
+}
+
+static int *open_pidfds(size_t n_pids, pid_t *pids)
+{
+	int *pidfds = calloc(n_pids, sizeof(*pidfds));
+	if (!pidfds)
+		return NULL;
+
+	for (size_t i = 0; i < n_pids; i++) {
+		pidfds[i] = pidfd_open(pids[i], 0);
+		if (pidfds[i] == -1) {
+			warn("could not open pid %d", pids[i]);
+			for (size_t j = 0; j < i; j++)
+				close(pidfds[j]);
+			goto fail;
+		}
+	}
+
+	return pidfds;
+
+fail:
+	free(pidfds);
+	return NULL;
+}
+
+static int add_listeners(int epll, size_t n_pids, int * const pidfds)
+{
+	for (size_t i = 0; i < n_pids; i++) {
+		struct epoll_event evt = {
+			.events = EPOLLIN,
+			.data = { .u64 = i },
+		};
+		int ret = epoll_ctl(epll, EPOLL_CTL_ADD, pidfds[i], &evt);
+		if (ret)
+			return ret;
+	}
+	return 0;
+}
+
+static int wait_for_exits(int epll, size_t n_pids, pid_t * const pids, int * const pidfds)
+{
+	size_t n_left = n_pids;
+	while (n_left) {
+		struct epoll_event evt;
+		int ret = epoll_wait(epll, &evt, 1, -1);
+		if (ret == -1) {
+			if (errno == EINTR)
+				continue;
+			else
+				return -errno;
+		}
+		if (verbose)
+			fprintf(stderr, "PID %d finished\n", pids[evt.data.u64]);
+		assert((size_t) ret <= n_left);
+		int fd = pidfds[evt.data.u64];
+		epoll_ctl(epll, EPOLL_CTL_DEL, fd, NULL);
+		close(fd);
+		pidfds[evt.data.u64] = -1;
+		n_left -= ret;
+	}
+	return 0;
+}
+
+static void __attribute__((__noreturn__)) usage(void)
+{
+	FILE *out = stdout;
+
+	fputs(USAGE_HEADER, out);
+	fprintf(out, _(" %s [options] pid...\n"), program_invocation_short_name);
+
+	fputs(USAGE_OPTIONS, out);
+	fputs(_(" -v, --verbose         be more verbose\n"), out);
+
+	fputs(USAGE_SEPARATOR, out);
+	fprintf(out, USAGE_HELP_OPTIONS(23));
+
+	fprintf(out, USAGE_MAN_TAIL("waitpid(1)"));
+
+	exit(EXIT_SUCCESS);
+}
+
+static int parse_options(int argc, char **argv)
+{
+	int c;
+	static const struct option longopts[] = {
+		{ "verbose", no_argument, NULL, 'v' },
+		{ "version", no_argument, NULL, 'V' },
+		{ "help",    no_argument, NULL, 'h' },
+		{ NULL, 0, NULL, 0 },
+	};
+
+	while ((c = getopt_long (argc, argv, "vVh", longopts, NULL)) != -1) {
+		switch (c) {
+		case 'v':
+			verbose = true;
+			break;
+		case 'V':
+			print_version(EXIT_SUCCESS);
+		case 'h':
+			usage();
+		default:
+			errtryhelp(EXIT_FAILURE);
+		}
+	}
+
+	return optind;
+}
+
+int main(int argc, char **argv)
+{
+	int ret;
+
+	setlocale(LC_ALL, "");
+	bindtextdomain(PACKAGE, LOCALEDIR);
+	textdomain(PACKAGE);
+
+	int pid_idx = parse_options(argc, argv);
+	const size_t n_pids = argc - pid_idx;
```

Comment:
@t-8ch, please, please, define variables at the beginning of the code block (function, if(), ... etc).